File: baixar-e-solicitar-xml-sefaz/src/sefaz_http/parser.py
"""
Parsers HTML para o portal SEFAZ via HTTP.
"""
from __future__ import annotations

import html
import re
from dataclasses import dataclass, field
from typing import Any
from urllib.parse import parse_qs, urljoin, urlparse

from bs4 import BeautifulSoup
from bs4.element import Tag


@dataclass(slots=True)
class HtmlForm:
    """Representa um formulario HTML do portal."""

    url: str
    action_url: str
    method: str
    fields: dict[str, list[str]] = field(default_factory=dict)
    select_options: dict[str, dict[str, str]] = field(default_factory=dict)
    submit_buttons: dict[str, str] = field(default_factory=dict)
    title: str | None = None
    raw_html: str = ""

    def first_value(self, name: str, default: str = "") -> str:
        values = self.fields.get(name)
        if not values:
            return default
        return values[0]

    def as_payload(self) -> list[tuple[str, str]]:
        payload: list[tuple[str, str]] = []
        for name, values in self.fields.items():
            for value in values:
                payload.append((name, value))
        return payload


@dataclass(slots=True)
class DownloadInfo:
    """Metadados de um arquivo disponivel para download."""

    url: str
    nm_arquivo: str
    situacao: str
    tipo_download: str
    dt_solicitacao: str | None = None
    file_path: str | None = None
    row_text: str = ""


@dataclass(slots=True)
class DownloadListingPage:
    """Pagina de listagem de downloads."""

    url: str
    current_page: int | None
    downloads: list[DownloadInfo]
    page_links: dict[int, str] = field(default_factory=dict)
    next_page_url: str | None = None
    new_request_url: str | None = None


_REDIRECT_PATTERNS = (
    re.compile(r"location\.href\s*=\s*['\"]([^'\"]+)['\"]", re.IGNORECASE),
    re.compile(r"window\.location(?:\.href)?\s*=\s*['\"]([^'\"]+)['\"]", re.IGNORECASE),
    re.compile(r"document\.location(?:\.href)?\s*=\s*['\"]([^'\"]+)['\"]", re.IGNORECASE),
)


def parse_js_redirect(base_url: str, html_text: str) -> str | None:
    """Extrai um redirecionamento JavaScript simples da resposta."""
    for pattern in _REDIRECT_PATTERNS:
        match = pattern.search(html_text)
        if match:
            return urljoin(base_url, html.unescape(match.group(1)))
    return None


def parse_portal_menu_link(base_url: str, html_text: str, link_text: str) -> str | None:
    """Encontra um link pelo texto exibido no portal."""
    soup = BeautifulSoup(html_text, "html.parser")
    link = soup.find("a", string=lambda text: text and link_text.lower() in text.lower())
    if not link or not link.get("href"):
        return None
    return urljoin(base_url, link["href"])


def parse_link_by_predicate(base_url: str, html_text: str, predicate) -> str | None:
    """Encontra um link pelo predicado informado."""
    soup = BeautifulSoup(html_text, "html.parser")
    for link in soup.find_all("a", href=True):
        if predicate(link):
            return urljoin(base_url, link["href"])
    return None


def parse_form(base_url: str, page_url: str, html_text: str, form_name: str | None = None) -> HtmlForm:
    """Extrai o primeiro formulario util da pagina."""
    soup = BeautifulSoup(html_text, "html.parser")
    forms = soup.find_all("form")
    target_form: Tag | None = None

    if form_name:
        target_form = soup.find("form", attrs={"name": form_name})

    if target_form is None:
        for form in forms:
            if form.find(["input", "select", "textarea"]):
                target_form = form
                break

    if target_form is None:
        raise ValueError("Nenhum formulario encontrado na pagina")

    fields: dict[str, list[str]] = {}
    select_options: dict[str, dict[str, str]] = {}
    submit_buttons: dict[str, str] = {}

    for control in target_form.find_all(["input", "select", "textarea"]):
        if not isinstance(control, Tag):
            continue

        name = control.get("name") or control.get("id")
        if not name:
            continue

        tag_name = control.name.lower()
        input_type = (control.get("type") or "").lower()

        if tag_name == "select":
            options: dict[str, str] = {}
            selected_values: list[str] = []
            for option in control.find_all("option"):
                value = option.get("value", "")
                text = option.get_text(" ", strip=True)
                options[text] = value
                if option.has_attr("selected"):
                    selected_values.append(value)
            fields[name] = selected_values or [""]
            select_options[name] = options
            continue

        if tag_name == "textarea":
            fields[name] = [control.get_text()]
            continue

        if input_type in {"radio", "checkbox"}:
            if control.has_attr("checked"):
                fields.setdefault(name, []).append(control.get("value", "on"))
            continue

        value = control.get("value", "")
        if input_type == "submit":
            submit_buttons[name] = value
            continue

        fields.setdefault(name, []).append(value)

    title = None
    title_node = soup.find("p")
    if title_node:
        title = title_node.get_text(" ", strip=True) or None

    action = target_form.get("action") or page_url
    method = (target_form.get("method") or "GET").upper()

    return HtmlForm(
        url=page_url,
        action_url=urljoin(base_url, action),
        method=method,
        fields=fields,
        select_options=select_options,
        submit_buttons=submit_buttons,
        title=title,
        raw_html=html_text,
    )


def parse_error_message(html_text: str) -> str | None:
    """Extrai mensagens de erro/aviso exibidas pelo portal."""
    soup = BeautifulSoup(html_text, "html.parser")
    node = soup.select_one(".fontMessageError")
    if node:
        text = node.get_text(" ", strip=True)
        if text:
            return text

    text = soup.get_text(" ", strip=True)
    lowered = text.lower()
    markers = (
        "arquivo não foi localizado",
        "o arquivo não foi localizado",
        "nao foi localizado",
        "usuário ou senha inválidos",
        "usuario ou senha invalido",
    )
    for marker in markers:
        if marker in lowered:
            return text
    return None


def is_session_expired(html_text: str) -> bool:
    """Detecta respostas que voltaram para login/sessao expirada."""
    soup = BeautifulSoup(html_text, "html.parser")
    if soup.find("input", attrs={"name": "UserName"}) and soup.find("input", attrs={"name": "Password"}):
        return True
    text = soup.get_text(" ", strip=True).lower()
    return "atoacessocontabilista" in html_text.lower() or "acesso rápido" in text or "acesso rapido" in text


def parse_download_listing(base_url: str, page_url: str, html_text: str, ready_only: bool = False) -> DownloadListingPage:
    """Parseia a listagem de downloads da transacao T10461."""
    soup = BeautifulSoup(html_text, "html.parser")
    downloads: list[DownloadInfo] = []

    for link in soup.find_all("a", href=True):
        href = link.get("href", "")
        if "process.jsp" not in href or "nmArquivo=" not in href:
            continue

        absolute_url = urljoin(base_url, href)
        query = parse_qs(urlparse(absolute_url).query)
        nm_arquivo = _first_qs_value(query, "nmArquivo")
        tipo_download = _first_qs_value(query, "tdb_dsTipoDownload", "DESCONHECIDO")
        dt_solicitacao = _first_qs_value(query, "dtSolicitacao") or None
        file_path = _first_qs_value(query, "FilePath") or None
        situacao = _first_qs_value(query, "dsSituacao")

        row = link.find_parent("tr")
        row_text = row.get_text(" ", strip=True) if row else link.get_text(" ", strip=True)

        if not situacao and row:
            cells = [cell.get_text(" ", strip=True) for cell in row.find_all("td")]
            situacao = next(
                (cell for cell in cells if "PRONTO" in cell.upper() or "PROCESSANDO" in cell.upper() or "NENHUM" in cell.upper()),
                "",
            )

        if ready_only and "PRONTO" not in html.unescape(situacao or "").upper():
            continue

        downloads.append(
            DownloadInfo(
                url=absolute_url,
                nm_arquivo=html.unescape(nm_arquivo),
                situacao=html.unescape(situacao),
                tipo_download=html.unescape(tipo_download or "DESCONHECIDO"),
                dt_solicitacao=dt_solicitacao,
                file_path=file_path,
                row_text=row_text,
            )
        )

    page_links: dict[int, str] = {}
    current_page = None
    next_page_url = None
    new_request_url = None

    current_node = soup.select_one("td.pgAtualNav b, font.fontPgAtualNav b")
    if current_node:
        current_text = current_node.get_text(strip=True)
        if current_text.isdigit():
            current_page = int(current_text)

    for link in soup.find_all("a", href=True):
        text = link.get_text(" ", strip=True)
        absolute_url = urljoin(base_url, link["href"])
        if text.isdigit():
            page_links[int(text)] = absolute_url
        elif "Próximo" in text or "Prximo" in text or "Próximo" in absolute_url:
            next_page_url = absolute_url

        href = link.get("href", "")
        if "TransId=T10464" in href:
            new_request_url = absolute_url

    if new_request_url is None:
        new_request_url = parse_link_by_predicate(
            base_url,
            html_text,
            lambda link: "TransId=T10464" in (link.get("href") or "") or "novo" in link.get_text(" ", strip=True).lower(),
        )

    return DownloadListingPage(
        url=page_url,
        current_page=current_page,
        downloads=downloads,
        page_links=page_links,
        next_page_url=next_page_url,
        new_request_url=new_request_url,
    )


def _first_qs_value(query: dict[str, list[str]], key: str, default: str = "") -> str:
    values = query.get(key) or []
    if not values:
        return default
    return values[0]


def simplify_form_payload(form: HtmlForm, overrides: dict[str, Any], submit_name: str = "okButton") -> list[tuple[str, str]]:
    """Monta payload preservando a ordem e duplicidades do formulario original."""
    payload = form.as_payload()
    consumed_singletons: set[str] = set()
    result: list[tuple[str, str]] = []

    def next_override_value(name: str, original: str) -> str:
        value = overrides.get(name)
        if value is None:
            return original
        if isinstance(value, list):
            current = value.pop(0) if value else original
            return str(current)
        if name in consumed_singletons:
            return original
        consumed_singletons.add(name)
        return str(value)

    for name, value in payload:
        if name in form.submit_buttons:
            continue
        result.append((name, next_override_value(name, value)))

    submit_value = form.submit_buttons.get(submit_name)
    if submit_value is not None:
        result.append((submit_name, submit_value))
    elif submit_name in overrides:
        result.append((submit_name, str(overrides[submit_name])))

    for name, value in overrides.items():
        if name in form.fields or name in form.submit_buttons:
            continue
        if isinstance(value, list):
            for item in value:
                result.append((name, str(item)))
        else:
            result.append((name, str(value)))

    return result

File: baixar-e-solicitar-xml-sefaz/TS/src-ts/sefaz-http/parser.ts
import { load, type CheerioAPI } from "cheerio";
import type { Element } from "domhandler";
import type { DownloadInfo, DownloadListingPage, HtmlForm } from "../types.js";

const redirectPatterns = [
  /location\.href\s*=\s*['"]([^'"]+)['"]/i,
  /window\.location(?:\.href)?\s*=\s*['"]([^'"]+)['"]/i,
  /document\.location(?:\.href)?\s*=\s*['"]([^'"]+)['"]/i,
];

function urlJoin(baseUrl: string, href: string): string {
  return new URL(href, baseUrl).toString();
}

function firstQueryValue(url: string, key: string, defaultValue = ""): string {
  return new URL(url).searchParams.get(key) ?? defaultValue;
}

function textOf($: CheerioAPI, element: Element): string {
  return $(element).text().replace(/\s+/g, " ").trim();
}

function normalizeSearchText(value: string): string {
  return value
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/\s+/g, " ")
    .trim()
    .toLowerCase();
}

function decodedUrlVariants(value: string): string[] {
  const variants = [value.replace(/&amp;/gi, "&")];
  for (let index = 0; index < 2; index += 1) {
    const current = variants[variants.length - 1];
    if (current == null) {
      break;
    }
    try {
      const decoded = decodeURIComponent(current);
      if (decoded === current) {
        break;
      }
      variants.push(decoded);
    } catch {
      break;
    }
  }
  return variants;
}

function isNewRequestUrl(href: string): boolean {
  return decodedUrlVariants(href).some((variant) =>
    /(?:^|[?&;])transid\s*=\s*t10464(?:$|[&#;])/i.test(variant),
  );
}

function elementLabel($: CheerioAPI, element: Element): string {
  const root = $(element);
  const parts = [textOf($, element)];
  for (const attr of ["title", "alt", "value", "aria-label"]) {
    const value = root.attr(attr);
    if (value) {
      parts.push(value);
    }
  }
  root.find("[title],[alt],[value],[aria-label]").each((_, child) => {
    const childNode = $(child);
    for (const attr of ["title", "alt", "value", "aria-label"]) {
      const value = childNode.attr(attr);
      if (value) {
        parts.push(value);
      }
    }
  });
  return parts.join(" ");
}

function isNewRequestLabel(label: string): boolean {
  const normalized = normalizeSearchText(label);
  return /\bnov[oa]\b/.test(normalized) || normalized.includes("nova solicitacao");
}

export function parseJsRedirect(baseUrl: string, htmlText: string): string | undefined {
  for (const pattern of redirectPatterns) {
    const match = pattern.exec(htmlText);
    if (match?.[1]) {
      return urlJoin(baseUrl, match[1]);
    }
  }
  return undefined;
}

export function parsePortalMenuLink(baseUrl: string, htmlText: string, linkText: string): string | undefined {
  const $ = load(htmlText);
  const needle = linkText.toLowerCase();
  for (const element of $("a[href]").toArray()) {
    if (textOf($, element).toLowerCase().includes(needle)) {
      return urlJoin(baseUrl, $(element).attr("href") ?? "");
    }
  }
  return undefined;
}

export function parseLinkByPredicate(
  baseUrl: string,
  htmlText: string,
  predicate: ($: CheerioAPI, element: Element) => boolean,
): string | undefined {
  const $ = load(htmlText);
  for (const element of $("a[href]").toArray()) {
    if (predicate($, element)) {
      return urlJoin(baseUrl, $(element).attr("href") ?? "");
    }
  }
  return undefined;
}

export function firstValue(form: HtmlForm, name: string, defaultValue = ""): string {
  return form.fields[name]?.[0] ?? defaultValue;
}

export function formAsPayload(form: HtmlForm): Array<[string, string]> {
  const payload: Array<[string, string]> = [];
  for (const [name, values] of Object.entries(form.fields)) {
    for (const value of values) {
      payload.push([name, value]);
    }
  }
  return payload;
}

export function parseForm(baseUrl: string, pageUrl: string, htmlText: string, formName?: string): HtmlForm {
  const $ = load(htmlText);
  let target = formName ? $(`form[name="${formName}"]`).first() : undefined;
  if (!target?.length) {
    target = $("form")
      .filter((_, form) => $(form).find("input,select,textarea").length > 0)
      .first();
  }
  if (!target.length) {
    throw new Error("Nenhum formulario encontrado na pagina");
  }

  const fields: Record<string, string[]> = {};
  const selectOptions: Record<string, Record<string, string>> = {};
  const submitButtons: Record<string, string> = {};

  target.find("input,select,textarea").each((_, control) => {
    const node = $(control);
    const name = node.attr("name") || node.attr("id");
    if (!name) {
      return;
    }
    const tagName = control.tagName.toLowerCase();
    const inputType = (node.attr("type") ?? "").toLowerCase();

    if (tagName === "select") {
      const options: Record<string, string> = {};
      const selectedValues: string[] = [];
      node.find("option").each((__, option) => {
        const optionNode = $(option);
        const value = optionNode.attr("value") ?? "";
        const optionText = optionNode.text().replace(/\s+/g, " ").trim();
        options[optionText] = value;
        if (optionNode.attr("selected") !== undefined) {
          selectedValues.push(value);
        }
      });
      fields[name] = selectedValues.length ? selectedValues : [""];
      selectOptions[name] = options;
      return;
    }

    if (tagName === "textarea") {
      fields[name] = [node.text()];
      return;
    }

    if (inputType === "radio" || inputType === "checkbox") {
      if (node.attr("checked") !== undefined) {
        fields[name] = [...(fields[name] ?? []), node.attr("value") ?? "on"];
      }
      return;
    }

    const value = node.attr("value") ?? "";
    if (inputType === "submit") {
      submitButtons[name] = value;
      return;
    }
    fields[name] = [...(fields[name] ?? []), value];
  });

  const action = target.attr("action") || pageUrl;
  const method = ((target.attr("method") || "GET").toUpperCase() === "POST" ? "POST" : "GET") as "GET" | "POST";
  const title = $("p").first().text().replace(/\s+/g, " ").trim() || undefined;

  return {
    url: pageUrl,
    actionUrl: urlJoin(baseUrl, action),
    method,
    fields,
    selectOptions,
    submitButtons,
    title,
    rawHtml: htmlText,
  };
}

export function parseErrorMessage(htmlText: string): string | undefined {
  const $ = load(htmlText);
  const errorText = $(".fontMessageError").first().text().replace(/\s+/g, " ").trim();
  if (errorText) {
    return errorText;
  }

  const text = $.text().replace(/\s+/g, " ").trim();
  const lowered = text.toLowerCase();
  const markers = [
    "arquivo não foi localizado",
    "o arquivo não foi localizado",
    "nao foi localizado",
    "usuário ou senha inválidos",
    "usuario ou senha invalido",
  ];
  return markers.some((marker) => lowered.includes(marker)) ? text : undefined;
}

export function isSessionExpired(htmlText: string): boolean {
  const $ = load(htmlText);
  if ($('input[name="UserName"]').length && $('input[name="Password"]').length) {
    return true;
  }
  const text = $.text().replace(/\s+/g, " ").trim().toLowerCase();
  return htmlText.toLowerCase().includes("atoacessocontabilista") || text.includes("acesso rápido") || text.includes("acesso rapido");
}

export function parseDownloadListing(
  baseUrl: string,
  pageUrl: string,
  htmlText: string,
  readyOnly = false,
): DownloadListingPage {
  const $ = load(htmlText);
  const downloads: DownloadInfo[] = [];

  for (const element of $("a[href]").toArray()) {
    const href = $(element).attr("href") ?? "";
    if (!href.includes("process.jsp") || !href.includes("nmArquivo=")) {
      continue;
    }

    const absoluteUrl = urlJoin(baseUrl, href);
    let situacao = firstQueryValue(absoluteUrl, "dsSituacao");
    const row = $(element).parents("tr").first();
    const rowText = row.length ? row.text().replace(/\s+/g, " ").trim() : textOf($, element);

    if (!situacao && row.length) {
      const cells = row
        .find("td")
        .toArray()
        .map((cell) => $(cell).text().replace(/\s+/g, " ").trim());
      situacao = cells.find((cell) => /PRONTO|PROCESSANDO|NENHUM/i.test(cell)) ?? "";
    }

    if (readyOnly && !situacao.toUpperCase().includes("PRONTO")) {
      continue;
    }

    downloads.push({
      url: absoluteUrl,
      nmArquivo: firstQueryValue(absoluteUrl, "nmArquivo"),
      situacao,
      tipoDownload: firstQueryValue(absoluteUrl, "tdb_dsTipoDownload", "DESCONHECIDO") || "DESCONHECIDO",
      dtSolicitacao: firstQueryValue(absoluteUrl, "dtSolicitacao") || undefined,
      filePath: firstQueryValue(absoluteUrl, "FilePath") || undefined,
      rowText,
    });
  }

  const pageLinks: Record<number, string> = {};
  let currentPage: number | undefined;
  let nextPageUrl: string | undefined;
  let newRequestUrl: string | undefined;
  const linkCount = $("a[href]").length;

  const currentText = $("td.pgAtualNav b, font.fontPgAtualNav b").first().text().trim();
  if (/^\d+$/.test(currentText)) {
    currentPage = Number(currentText);
  }

  for (const element of $("a[href]").toArray()) {
    const text = textOf($, element);
    const href = $(element).attr("href") ?? "";
    const absoluteUrl = urlJoin(baseUrl, href);
    if (/^\d+$/.test(text)) {
      pageLinks[Number(text)] = absoluteUrl;
    } else if (text.includes("Próximo") || text.includes("Prximo") || absoluteUrl.includes("Próximo")) {
      nextPageUrl = absoluteUrl;
    }
    if (isNewRequestUrl(href) || isNewRequestLabel(elementLabel($, element))) {
      newRequestUrl = absoluteUrl;
    }
  }

  if (!newRequestUrl) {
    newRequestUrl = parseLinkByPredicate(
      baseUrl,
      htmlText,
      ($api, element) =>
        isNewRequestUrl($api(element).attr("href") ?? "") ||
        isNewRequestLabel(elementLabel($api, element)),
    );
  }

  return { url: pageUrl, currentPage, downloads, pageLinks, nextPageUrl, newRequestUrl, linkCount };
}

export function parseEmpresasDoFormulario(form: HtmlForm): Array<{ inscricao: string; nome: string }> {
  const empresas: Array<{ inscricao: string; nome: string }> = [];
  for (const [texto, valor] of Object.entries(form.selectOptions.cdPessoaContribuinte ?? {})) {
    const inscricao = valor.trim();
    const nome = texto.trim();
    if (!inscricao || !nome || nome.toLowerCase().startsWith("selecione")) {
      continue;
    }
    empresas.push({ inscricao, nome });
  }
  return empresas;
}

export function simplifyFormPayload(
  form: HtmlForm,
  overrides: Record<string, unknown>,
  submitName = "okButton",
): Array<[string, string]> {
  const consumedSingletons = new Set<string>();
  const result: Array<[string, string]> = [];
  const mutableOverrides = new Map(Object.entries(overrides));

  function nextOverrideValue(name: string, original: string): string {
    const value = mutableOverrides.get(name);
    if (value == null) {
      return original;
    }
    if (Array.isArray(value)) {
      const current = value.length ? value.shift() : original;
      return String(current);
    }
    if (consumedSingletons.has(name)) {
      return original;
    }
    consumedSingletons.add(name);
    return String(value);
  }

  for (const [name, value] of formAsPayload(form)) {
    if (name in form.submitButtons) {
      continue;
    }
    result.push([name, nextOverrideValue(name, value)]);
  }

  const submitValue = form.submitButtons[submitName];
  if (submitValue != null) {
    result.push([submitName, submitValue]);
  } else if (submitName in overrides) {
    result.push([submitName, String(overrides[submitName])]);
  }

  for (const [name, value] of Object.entries(overrides)) {
    if (name in form.fields || name in form.submitButtons) {
      continue;
    }
    if (Array.isArray(value)) {
      for (const item of value) {
        result.push([name, String(item)]);
      }
    } else {
      result.push([name, String(value)]);
    }
  }

  return result;
}
